Collected fixes for open issues

gh-mlfowler opened the pull request:
Having made heavy use of your tool on a recent project I have encountered a number of problems which I have resolved on my own fork. I have collected these fixes into a single PR and raised corresponding issues for them.

- Resolve #2 - In commit 8c5e75f I have added the RedShift parameter to limit the size of the unloaded CSV updating the documentation in commit e28e61d 
- Resolve #14 - Reworked the AWS credentials so that either the file or environment can be used in commit 1f7ed0a
- Resolve #15 - Simplified the error handling and simply output location if present instead of attempting to parse it in commit 0c19085 
- Resolve #16 - Removed the bespoke timestamp handling. The default output from RedShift is correctly parsed by BigQuery. Commit bd4916f 
- Resolve #17 - Rewrote the schema query to extract the column positions to preserve them for BigQuery in commit 707df6c 
- Resolve #18 - Manipulated the raw time to schedule in the future to avoid wraparound in commit 5a98691 
- Resolve #19 - Specified the BigQuery option  `allow_quoted_newlines` in commit c48182e 
- Resolve #20 - Reworked GCP credential handling to use the environment (commit 
6f436df) or the compute engine scopes (commit 71fa5a2)
- Resolve #21 - Forced invocations of RedShift to always use a new connection in commit 
d3da6af

gh-mlfowler committed:
Preserve column ordering

gh-mlfowler committed:
Manipulate raw time to schedule in the future

Manipulating the minute value caused the 59th minute to become 60 which
is rejected by the Google API

```Invalid minutes specified. Minutes must be between 0 and 59
inclusive.```

gh-mlfowler committed:
Simplify error reporting

When `error.location` doesn't contain a string that matches the expected
output a `nil` is encountered causing a new error to propagate losing
the one we're interested in. Instead simply report the value of
`error.location` if it is present.

gh-mlfowler committed:
Redshift strings can contain newlines

gh-mlfowler committed:
Remove bespoke TIMESTAMP handling

Using epoch dates prevents some dates from being unloaded. The default
string conversion of TIMESTAMP is importable into BigQuery anyway.

gh-mlfowler committed:
Set max UNLOAD file size to 3.9GB

BigQuery has a max file size of 4GB whereas RedShift has 6.2GB. Setting
RedShift to 4GB also generated a file slightly too large for BigQuery.

gh-mlfowler committed:
Additional logging to indicate progress

gh-mlfowler committed:
Add support for using GOOGLE_APPLICATION_CREDENTIALS

This allows for the credentials to be specified externally to the
command line.

gh-mlfowler committed:
Rework GCP authorization to use service account

Allow the Google SDK to use the instance metadata to obtain the
service account if Bigshift is running within GCP.

gh-mlfowler committed:
Remove size troubleshoot

The issue was resolved with commit `8c5e75f`

gh-mlfowler committed:
Update manual credential handling

Using the CredentialProviderChain with a manually created Credentials
object causes internal errors however the object works with the API.
Therefore only use the CredentialProviderChain if --aws-credentials
isn't specified to extract the configuration from the environment.

gh-mlfowler committed:
Use fresh connections for Redshift

Consistently seeing this error when the schema is queried for a load:
```PG::ConnectionBad: PQconsumeInput() SSL error: decryption failed or bad record mac

  /home/mlfowler/git/bigshift/lib/bigshift/cli.rb:237:in `async_exec'
  /home/mlfowler/git/bigshift/lib/bigshift/cli.rb:237:in
`rs_connection'```

gh-mlfowler committed:
Update tests to match recent changes

iconara commented on lib/bigshift/redshift_table_schema.rb:
```diff
@@ -8,7 +8,17 @@ def initialize(schema_name, table_name, redshift_connection)
+          AND ptd.column = isc.column_name
+          AND "schemaname" = $1
+          AND "tablename" = $2
+          ORDER BY ordinal_position
```
I think either all table and column names should be quoted, or none should be. The pieces you've added are not quoted. I'm fine with either and since this SQL is not generated and the names are not mixed case, there is no need for quotes really. I would like it to be consistent, though, so do either or, but not both.

gh-mlfowler replied:
In the end I have gone for unquoted with the exception of the columns in the select as `column` is a keyword.

iconara replied:
👌

iconara commented on lib/bigshift/redshift_table_schema.rb:
```diff
@@ -8,7 +8,17 @@ def initialize(schema_name, table_name, redshift_connection)
+          AND "schemaname" = $1
+          AND "tablename" = $2
+          ORDER BY ordinal_position
+        }.gsub(/\s+/, " ").strip
```
The space could be single quoted to show that no string interpolation is necessary.

iconara commented on lib/bigshift/cli.rb:
```diff
@@ -268,15 +278,27 @@ def aws_region
+        elsif (credentials = Google::Auth::GCECredentials.new)
+          credentials
+        else
+          raise 'No GCP credentials found'
```
This should raise `CliError` like in other places in this file.

iconara replied:
I just saw I had missed this too when validating the AWS credentials. I don't remember if that was intentional or not. Either ignore the above comment, or change both instances.

gh-mlfowler replied:
It turns out that in the absence of the credential files the other mechanism for both AWS and GCP will always create an object however you don't know if it's valid until you attempt to invoke the relevant API. After a bunch of testing I added some top level exception handling to tell the user their error in https://github.com/iconara/bigshift/blob/e0bc51be55652338231345edb7cada960e766f92/lib/bigshift/cli.rb#L27-L44

iconara replied:
This looks like a brittle solution, unfortunately. It's almost always wrong to catch `NoMethodError` (and "match" doesn't seem like a very specific filter).

Could you explain a bit more why this happens (the `NoMethodError`) and perhaps I can help you figure out how to make it less brittle.

gh-mlfowler replied:
The error occurs if you don't specify the AWS credentials either in the environment or as a file. Google's SDK raises a specific exception but the AWS library sadly doesn't. I could also check the trace but that would make it even more brittle. I was trying to avoid writing an AWS configuration validation but perhaps that's the best thing to do? 

```AWS configuration missing or malformed: undefined method `match' for nil:NilClass:                                                                                                                                                                  
                                                                                                                                                                                                                                                    
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:82:in `block in partition_matching_region'                                                                           
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:81:in `each'                                                                                                         
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:81:in `find'                                                                                                         
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:81:in `partition_matching_region'                                                                                    
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:70:in `get_partition'                                                                                                
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:23:in `signing_region'                                                                                               
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:99:in `signing_region'                                                                                               
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-s3-1.29.0/lib/aws-sdk-s3/plugins/s3_signer.rb:20:in `block in <class:S3Signer>'                                                                                             
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:70:in `call'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:211:in `block in resolve_defaults'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:57:in `each'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:57:in `each'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:210:in `resolve_defaults'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:203:in `value_at'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:189:in `block in resolve'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/2.5.0/set.rb:338:in `each_key'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/2.5.0/set.rb:338:in `each'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:189:in `resolve'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:177:in `apply_defaults'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:150:in `build!'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/base.rb:62:in `build_config'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/base.rb:19:in `initialize'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-s3-1.29.0/lib/aws-sdk-s3/client.rb:262:in `initialize'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/base.rb:99:in `new'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-s3-1.29.0/lib/aws-sdk-s3/resource.rb:14:in `initialize'
* /home/~/git/bigshift/lib/bigshift/cli.rb:218:in `new'
* /home/~/git/bigshift/lib/bigshift/cli.rb:218:in `s3_resource'
* /home/~/git/bigshift/lib/bigshift/cli.rb:229:in `create_unload_manifest'
* /home/~/git/bigshift/lib/bigshift/cli.rb:68:in `unload'
* /home/~/git/bigshift/lib/bigshift/cli.rb:30:in `run'
* /home/~/git/bigshift/bin/bigshift:6:in `<top (required)>'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/bin/bigshift:23:in `load'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/bin/bigshift:23:in `<top (required)>'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `load'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `kernel_load'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:28:in `run'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:463:in `exec'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:27:in `dispatch'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:18:in `start'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/exe/bundle:30:in `block in <top (required)>'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/exe/bundle:22:in `<top (required)>'
* /home/~/.rbenv/versions/2.5.3/bin/bundle:23:in `load'
* /home/~/.rbenv/versions/2.5.3/bin/bundle:23:in `<main>'```

iconara commented on lib/bigshift/cli.rb:
```diff
@@ -253,10 +266,7 @@ def aws_credentials
-          credentials = nil
-        end
-        if (credentials = Aws::CredentialProviderChain.new(credentials).resolve)
+        elsif (credentials = Aws::CredentialProviderChain.new().resolve)
```
There is no need for `()` here.

iconara commented on lib/bigshift/cli.rb:
```diff
@@ -268,15 +278,27 @@ def aws_region
+            json_key_io: StringIO.new(JSON.dump(@config[:gcp_credentials])),
+            scope: Google::Apis::StoragetransferV1::AUTH_CLOUD_PLATFORM
+          )
+        elsif (credentials = Google::Auth::GCECredentials.new)
```
Can `.new` really return `nil` here? I know it's technically possible to do, but it's very unusual.

iconara commented on lib/bigshift.rb:
```diff
@@ -1,6 +1,7 @@
 require 'google/apis/bigquery_v2'
 require 'google/apis/storagetransfer_v1'
 require 'google/apis/storage_v1'
+require "google/cloud/env"
```
Use single quotes here, like on the other lines.

iconara reviewed (commented):
Thank you for helping out getting this updated. It was a long time ago that I used it myself, so I'm not surprised there were things that no longer were working.

Your changes look good overall, but I have some changes I'd like you to make, just to keep the code and tests consistent.

At first I only saw your issues and I thought that I'd have to give you the bad news that I would probably never fix them, so I was pleasantly surprised to see your PR with all the fixes. It's much appreciated.

gh-mlfowler commented:
Thanks for the feedback @iconara! I haven't done a great deal of Ruby so your code review is greatly appreciated. I had hoped to have the changes done by now but it looks like I won't have time until early next week, but they will come.

gh-mlfowler committed:
Refinements based on PR review

gh-mlfowler committed:
Re-work these tests having re-read the feedback

gh-mlfowler commented:
For the most part I have taken your suggestions as is and updated my code @iconara so I have resolved the comments. In a couple of cases I've done a bit more than the obvious so have left those comments open for you to take another look at.

iconara commented on lib/bigshift/cli.rb:
```diff
@@ -25,12 +25,22 @@ def initialize(argv, options={})
+        if ['access_key_id', 'match'].any? { |key| e.message.include? key }
+          raise CliError.new('AWS configuration missing or malformed: ' + e.message, e.backtrace, @usage)
+        else
+          raise CliError.new(e.message, e.backtrace, @usage)
```
I mentioned in another comment (it was a thread from the last review, it doesn't show in the diff any longer, unfortunately) that I would like this to not catch `NoMethodError` – but if it turns out that it will have to do that for some reason, this branch should not translate all other `NoMethodError`s to `CliError`. This should just `raise`. The other branch catches an expected error and translates it into a more semantic error, but this branch is for the unexpected errors and they should not be masked as something which they aren't.

iconara reviewed (approved):
There are a few small things that I'd like changed before this is merged, but overall this looks really good now.

gh-mlfowler committed:
Propagate the unexpected exception rather than wrap it

gh-mlfowler commented:
I have propagated the unexpected exception rather than wrap it in CliError and have explained my thinking around the AWS credentials exception handling @iconara.

iconara commented:
Thank you @gh-mlfowler for the update.

I think the AWS SDK error is a bug in aws-sdk-s3: `Aws::S3::Client.new` raises `NoMethodError`, but `Aws::S3::Client.new(region: 'foo')` raises `Aws::Sigv4::Errors::MissingCredentialsError`. That should probably be reported, but either way I think the problem at hand should be solved by checking the region argument before creating the client. If the AWS SDK fixes the bug the `NoMethodError` check will be broken and the `Aws::Errors::MissingRegionError` that should be raised will bubble up to the top level.

My suggestion is either to report the bug and ignore the `NoMethodError` (in the hope that it will go away by itself when aws-sdk-s3 is updated), or to add validation in `#aws_region` that raises a specific error that can be rescued and translated into a `CliError` in `#run`. In both scenarios `Aws::Errors::MissingRegionError` and `Aws::Sigv4::Errors::MissingCredentialsError` should be rescued in `#run` and translated into a `CliError` explaining what was missing.

gh-mlfowler committed:
Handle missing AWS credentials as documented.

The NoSuchMethod error appears to be a bug in the SDK as per @iconara's
investigation. Re-worked the exception handling on the assumption of
correct SDK behaviour and documneted it as a known issue.

gh-mlfowler commented:
@iconara I have re-worked the exception handling to match the expected SDK behaviour and added a known issue to the readme. Apologies I've not got back to this for so long!

iconara reviewed (approved):
Looks good. I'll merge and push a new release as soon as I can.

iconara commented:
Released as [v0.4.0](https://github.com/iconara/bigshift/releases/tag/v0.4.0). I also changed from depending on `aws-sdk` to `aws-sdk-s3` to cut down on the dependency tree a bit.
